src/herder/TxSetFrame.cpp

```diff
@@ -186,21 +186,23 @@ TxSetFrame::surgePricingFilter(LedgerManager const& lm)
     }
 }
 
-// TODO.3 this and checkValid share a lot of code
-void
-TxSetFrame::trimInvalid(Application& app,
-                        std::vector<TransactionFramePtr>& trimmed)
+template <typename Lambda1, typename Lambda2, typename Lambda3>
+bool
+TxSetFrame::checkOrTrim(Application &app,
+                        Lambda1 checkTxLambda,
+                        Lambda2 processInvalidTxLambda,
+                        Lambda3 processLastInvalidTxLambda)
 {
-    soci::transaction sqltx(app.getDatabase().getSession());
-    app.getDatabase().setCurrentTransactionReadOnly();
-
-    sortForHash();
-
     map<AccountID, vector<TransactionFramePtr>> accountTxMap;
 
-    for (auto tx : mTransactions)
+    Hash lastHash;
+    for (auto& tx : mTransactions)
     {
+        if (!checkTxLambda(tx, lastHash)) {
```

Comment:
You can ditch the checkTxLambda and always check: it will add a very small overhead to the "Trim" path but the code ends up being simpler  and "Trim" is not called that often